updated method to get topics for recent_blog_post filter

danielfmiranda opened the pull request:
# Description

<!-- Describe the PR here -->

Link to sample test page:
Related PRs/issues: #8224 

The `RecentBlogEntries` block used to have a hard coded list of topics that were able to be used in it's "topic filter":

This PR introduces a minor change to the method that the topic filter uses in order to get its list of possible options, as well as the necessary migrations for the change. Now, the list is updated dynamically and includes any new topics that have been added to the DB.
 
# Checklist

<!-- Check off items with `[x]` or cross out items that don't apply with `~~The description~~` -->

**Tests**
- ~~[ ] Is the code I'm adding covered by tests?~~

**Changes in Models:**
- ~~[ ] Did I update or add new fake data?~~
- ~~[ ] Did I squash my migration?~~
- ~~[ ] Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?~~

**Documentation:**
- [x] Is my code documented?
- ~~[ ] Did I update the READMEs or wagtail documentation?~~



# Steps to test

1. Pull this branch and run `inv new-db`
2. Log into the CMS, and visit a page that can use the `RecentBlogEntries` block. For a quick link, you can visit a campaign page at http://localhost:8000/cms/pages/87/edit/
3. Add a new `RecentBlogEntries` block to the body, and select the drop-down for the "Filter By Topic" section. Note that there are only a handful of options, which are the topics that are currently in the DB.
<img width="2086" alt="image" src="https://user-images.githubusercontent.com/18314510/195947513-2893edc7-76cf-4336-a4b5-5e9fc8e4ffee.png">

4. Visit http://localhost:8000/cms/snippets/wagtailpages/blogpagetopic/add/?locale=en, and create a new Topic.

5. Go back to the `RecentBlockEntries` at http://localhost:8000/cms/pages/87/edit/, and take a look at the Topic options again. Note that the new topic is in the list.

<img width="2086" alt="Screen Shot 2022-10-14 at 2 36 52 PM" src="https://user-images.githubusercontent.com/18314510/195947902-4e8cea49-1aa3-4324-8a3f-63cf7e22185e.png">


6. If everything is working as expected, testing is complete!

tbrlpld commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_topic.py:
```diff
@@ -49,28 +49,16 @@ class BlogPageTopic(TranslatableMixin, models.Model):
+        # without this, it will return a ProgrammingError due to BlogPageTopics not yet existing.
+        try:
+            choices = [(cat.name, cat.name) for cat in BlogPageTopic.objects.all()]
+            choices.sort(key=lambda c: c[1])
```
Have you considered performing the sorting in the database? I think that might also allow you to avoid the `ProgrammingError` since I guess that happens during the sort right?

danielfmiranda replied:
Great idea Tibor, this has been updated!

tbrlpld commented on network-api/networkapi/wagtailpages/pagemodels/blog/blog_topic.py:
```diff
@@ -49,28 +49,16 @@ class BlogPageTopic(TranslatableMixin, models.Model):
 
     @staticmethod
     def get_topics():
```
Have you considered adding a test for this method?

tbrlpld replied:
Also, what do think about turning this into a `classmethod`? To me this looks like one, since it does not depend on the instance but does need the class itself.

danielfmiranda replied:
Hi Tibor, this has been updated to use a `classmethod` now!

tbrlpld reviewed (commented):
Thanks @danielfmiranda. This is coming a long nicely. I have left some small questions.

danielfmiranda commented:
Hi @tbrlpld thank you very much for your review and feedback. I have implemented your suggestions and also added a test for the BlogPageTopic models "get_topics" method. This PR is now ready for re-review whenever you have a free moment. Thanks again!

tbrlpld commented on network-api/networkapi/wagtailpages/tests/test_blog.py:
```diff
@@ -860,3 +860,27 @@ def test_authors_detail_non_existent_id_argument(self):
+        blog_factories.BlogPageTopicFactory(name="Test_Topic_2")
+        blog_factories.BlogPageTopicFactory(name="Test_Topic_3")
+
+        # Creating a list of all created BlogPageTopics sorted by name,
+        # with an additional option of "All".
+        list_of_sorted_topics = [('All', 'All')]
+        for topic in blog_models.BlogPageTopic.objects.all().order_by('name'):
+            list_of_sorted_topics.append((topic.name, topic.name))
```
What do you think of just spelling this out manually? I find it often a bit hard to understand a test when there is a lot of automation in the setup. Also, this looks basically like the implementation 😅

danielfmiranda replied:
Hi Tibor, after giving this a second look I agree that just having the list defined and shown in the order it should be in is a little more clear then doing the sorting and arranging this way, I have updated the test!

tbrlpld reviewed (approved):
Thanks @danielfmiranda. This looks very good to me. Thanks for adding the test. 

I have left one small question around the test setup, but that is up to you if you want to followup on that.
